src/main/host/syscall/poll.c

```diff
@@ -0,0 +1,241 @@
+        // Rely on the kernel to poll the OS-back file
+        int res = file_poll((File*)ldesc, pfd);
+        if (res < 0) {
+            warning("Asking the kernel to poll file %i resulted in error %i: %s", pfd->fd, -res,
+                    strerror(-res));
+            pfd->revents |= POLLERR;
+        }
+    } else {
+        // Figure out which events to report
+        if ((dstat & STATUS_DESCRIPTOR_CLOSED) && !(dstat & STATUS_DESCRIPTOR_ACTIVE)) {
+            pfd->revents |= POLLNVAL;
+        }
+        if ((pfd->events & POLLIN) && (dstat & STATUS_DESCRIPTOR_ACTIVE) &&
+            (dstat & STATUS_DESCRIPTOR_READABLE)) {
+            pfd->revents |= POLLIN;
+        }
+        if ((pfd->events & POLLOUT) && (dstat & STATUS_DESCRIPTOR_ACTIVE) &&
+            (dstat & STATUS_DESCRIPTOR_WRITABLE)) {
+            pfd->revents |= POLLOUT;
+        }
+    }
+}
+
+static int _syscallhandler_getPollEvents(SysCallHandler* sys, struct pollfd* fds, nfds_t nfds) {
+    int num_ready = 0;
+
+    for (nfds_t i = 0; i < nfds; i++) {
+        struct pollfd* pfd = &fds[i];
+        pfd->revents = 0;
+
+        // Negative fd just means skip over this one
+        if (pfd->fd < 0) {
+            // Continue is ok because we have no revents to count
+            continue;
+        }
+
+        debug("poll checking fd %i", pfd->fd);
+
+        /* Get the descriptor. */
+        CompatDescriptor* cdesc = process_getRegisteredCompatDescriptor(sys->process, pfd->fd);
+        if (cdesc) {
+            _syscallhandler_getPollEventsHelper(cdesc, pfd);
+        } else {
+            pfd->revents |= POLLNVAL;
+        }
+
+        // Count if we are reporting an event
+        num_ready += (pfd->revents == 0) ? 0 : 1;
+    }
+
+    return num_ready;
+}
+
+static void _syscallhandler_registerPollFDs(SysCallHandler* sys, struct pollfd* fds, nfds_t nfds) {
+    // Epoll should already be clear, but let's make sure
+    epoll_reset(sys->epoll);
+
+    for (nfds_t i = 0; i < nfds; i++) {
+        struct pollfd* pfd = &fds[i];
+
+        if (pfd->fd < 0) {
+            continue;
+        }
+
+        CompatDescriptor* cdesc = process_getRegisteredCompatDescriptor(sys->process, pfd->fd);
+        utility_assert(cdesc); // we would have returned POLLNVAL in getPollEvents
+
+        struct epoll_event epev = {0};
+        if (pfd->events & POLLIN) {
+            epev.events |= EPOLLIN;
+        }
+        if (pfd->events & POLLOUT) {
+            epev.events |= EPOLLOUT;
+        }
+
+        if (epev.events) {
+            epoll_control(sys->epoll, EPOLL_CTL_ADD, pfd->fd, cdesc, &epev);
+        }
+    }
+}
+
+static SysCallReturn _syscallhandler_pollHelper(SysCallHandler* sys, PluginPtr fds_ptr, nfds_t nfds,
+                                                const struct timespec* timeout) {
+    // Get the pollfd struct in our memory
+    struct pollfd* fds =
+        process_getWriteablePtr(sys->process, sys->thread, fds_ptr, nfds * sizeof(*fds));
```

Comment:
Should this be `process_getMutablePtr`?

Reply:
Oh yeah, whoops. Not sure how `process_getWriteablePtr` was working actually, because we do need to read AND write so `process_getMutablePtr` should be used. Good catch!